### PR TITLE
cata reforging and void storage

### PR DESCRIPTION
**Pull request by aaron02**
**Description**
Activated reforging in cata and optimized transmog visual in character screen
Activated void storage

**Todo / Checklist**
- [x] Target is branch develop.
- [x] Detailed description about this pull request.
- [x] Checked AE-Coding standards.

**Tests Performed:** 
- [x] Build AE.
- [x] Server startup.
- [x] Log into world.

***Multiversion Ingame Tests Performed:***
- [x] BC
- [x] WotLK
- [x] Cata

### CODE REVIEW

File: src/world/Management/Item.cpp [hunk 1]
```diff
@@ -330,3 +331,61 @@ bool Item::canTransmogrifyItemWithItem(Item const* transmogrified, Item const* t
...
+
+    if (randomPropId < 0)
+    {
+        DBC::Structures::ItemRandomSuffixEntry const* randomSuffix = sItemRandomSuffixStore.LookupEntry(-randomPropId);
+        if (!randomSuffix)
+            return 0;
+
+        for (uint32_t e = PROP_ENCHANTMENT_SLOT_0; e <= PROP_ENCHANTMENT_SLOT_4; ++e)
+            if (DBC::Structures::SpellItemEnchantmentEntry const* enchant = sSpellItemEnchantmentStore.LookupEntry(getEnchantmentId(EnchantmentSlot(e))))
+                for (uint32_t f = 0; f < MAX_ITEM_ENCHANTMENT_EFFECTS; ++f)
```
Comment (Appled) [inline src/world/Management/Item.cpp:368]: You can replace uint32_t with uint8_t here

File: src/world/Management/Item.cpp [hunk 2]
```diff
@@ -330,3 +331,61 @@ bool Item::canTransmogrifyItemWithItem(Item const* transmogrified, Item const* t
...
+    }
+    else
+    {
+        DBC::Structures::ItemRandomPropertiesEntry const* randomProp = sItemRandomPropertiesStore.LookupEntry(randomPropId);
+        if (!randomProp)
+            return 0;
+
+        for (uint32_t e = PROP_ENCHANTMENT_SLOT_0; e <= PROP_ENCHANTMENT_SLOT_4; ++e)
+            if (DBC::Structures::SpellItemEnchantmentEntry const* enchant = sSpellItemEnchantmentStore.LookupEntry(getEnchantmentId(EnchantmentSlot(e))))
+                for (uint32_t f = 0; f < MAX_ITEM_ENCHANTMENT_EFFECTS; ++f)
```
Comment (Appled) [inline src/world/Management/Item.cpp:382]: You can replace uint32_t with uint8_t here

File: src/world/Management/Item.cpp [hunk 3]
```diff
@@ -330,3 +331,61 @@ bool Item::canTransmogrifyItemWithItem(Item const* transmogrified, Item const* t
...
+                    if (enchant->type[f] == ITEM_ENCHANTMENT_TYPE_STAT && ItemModType(enchant->spell[f]) == statType)
+                        for (int k = 0; k < MAX_ITEM_ENCHANTMENT_EFFECTS; ++k)
```
Comment (Appled) [inline src/world/Management/Item.cpp:384]: You can replace int with uint8_t here

File: src/world/Units/Players/PlayerDefines.hpp [hunk 4]
```diff
@@ -589,32 +589,37 @@ const time_t forcedResurrectInterval = 360000;  // 1000*60*6=6 minutes
...
+    CR_HIT_RANGED                       = 6,
+    CR_HIT_SPELL                        = 7,
+    CR_CRIT_MELEE                       = 8,
+    CR_CRIT_RANGED                      = 9,
+    CR_CRIT_SPELL                       = 10,
+    CR_HIT_TAKEN_MELEE                  = 11,   // Deprecated since Cataclysm
+    CR_HIT_TAKEN_RANGED                 = 12,   // Deprecated since Cataclysm
+    CR_HIT_TAKEN_SPELL                  = 13,   // Deprecated since Cataclysm
+    CR_RESILIENCE_CRIT_TAKEN            = 14,
+    CR_RESILIENCE_PLAYER_DAMAGE_TAKEN   = 15,
```
Comment (Appled) [inline src/world/Units/Players/PlayerDefines.hpp:607]: This is no longer correct for versions < Cata (should be ranged crit resilience)

File: src/world/Units/Players/PlayerDefines.hpp [hunk 5]
```diff
@@ -589,32 +589,37 @@ const time_t forcedResurrectInterval = 360000;  // 1000*60*6=6 minutes
...
+    CR_HIT_MELEE                        = 5,
+    CR_HIT_RANGED                       = 6,
+    CR_HIT_SPELL                        = 7,
+    CR_CRIT_MELEE                       = 8,
+    CR_CRIT_RANGED                      = 9,
+    CR_CRIT_SPELL                       = 10,
+    CR_HIT_TAKEN_MELEE                  = 11,   // Deprecated since Cataclysm
+    CR_HIT_TAKEN_RANGED                 = 12,   // Deprecated since Cataclysm
+    CR_HIT_TAKEN_SPELL                  = 13,   // Deprecated since Cataclysm
+    CR_RESILIENCE_CRIT_TAKEN            = 14,
```
Comment (Appled) [inline src/world/Units/Players/PlayerDefines.hpp:606]: This is no longer correct for versions < Cata (should be melee crit resilience)

**Review by Appled — CHANGES_REQUESTED** [review]

File: src/world/GameCata/Storage/DBCStructures.h [hunk 6]
```diff
@@ -1239,11 +1240,13 @@ namespace DBC::Structures
...
     };
 
+#define MAX_ITEM_ENCHANTMENT_EFFECTS 3
+
     struct ItemRandomPropertiesEntry
     {
         uint32_t ID;                                                // 0
         //char* name1;                                              // 1
-        uint32_t spells[5];                                         // 2-6 enchant_id
+        uint32_t spells[MAX_ITEM_ENCHANTMENT_EFFECTS];                                         // 2-6 enchant_id
```
[the remark this answers is not in the document]
Comment (aaron02) [reply src/world/GameCata/Storage/DBCStructures.h:1249]: the other colums are always empty.  there are always no more than 3 enchatns available on an item

File: src/world/Units/Players/Player.cpp [hunk 7]
```diff
@@ -169,6 +172,11 @@ Player::~Player()
     delete SDetector;
     SDetector = nullptr;
 
+#if VERSION_STRING > WotLK
+    for (uint8 i = 0; i < VOID_STORAGE_MAX_SLOT; ++i)
```
Comment (Appled) [inline src/world/Units/Players/Player.cpp:176]: uint8_t

File: src/world/Server/Packets/Handlers/VoidStorageHandler.cpp [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+    }
+
+    uint8_t oldSlot;
+    if (!player->getVoidStorageItem(itemId, oldSlot))
+    {
+        sLogger.debug("handleVoidSwapItem - Player (GUID: %u, name: %s) requested swapping an invalid item (slot: %u, itemid: %I64u).", player->getGuidLow(), player->getName().c_str(), newSlot, uint64(itemId));
+        return;
+    }
+
+    bool usedSrcSlot = player->getVoidStorageItem(oldSlot) != NULL; // should be always true
```
Comment (Appled) [inline src/world/Server/Packets/Handlers/VoidStorageHandler.cpp:557]: Do not use NULL, always nullptr

File: src/world/Server/Packets/Handlers/VoidStorageHandler.cpp [hunk 9]
```diff
@@ -0,0 +1,638 @@
...
+    bool usedDestSlot = player->getVoidStorageItem(newSlot) != NULL;
```
Comment (Appled) [inline src/world/Server/Packets/Handlers/VoidStorageHandler.cpp:558]: Do not use NULL, always nullptr

File: src/world/Units/Players/Player.cpp [hunk 10]
```diff
@@ -4071,3 +4290,177 @@ bool Player::isOnGMTargetList(uint32_t guid) const
...
+uint8_t Player::getNextVoidStorageFreeSlot() const
+{
+    for (uint8_t i = 0; i < VOID_STORAGE_MAX_SLOT; ++i)
+        if (!_voidStorageItems[i]) // unused item
+            return i;
+
+    return VOID_STORAGE_MAX_SLOT;
+}
+
+uint8 Player::getNumOfVoidStorageFreeSlots() const
```
Comment (Appled) [inline src/world/Units/Players/Player.cpp:4303]: uint8_t

File: src/world/Units/Players/Player.cpp [hunk 11]
```diff
@@ -4071,3 +4290,177 @@ bool Player::isOnGMTargetList(uint32_t guid) const
...
+void Player::deleteVoidStorageItem(uint8_t slot)
+{
+    if (slot >= VOID_STORAGE_MAX_SLOT)
+    {
+        GetSession()->sendVoidStorageTransferResult(VOID_TRANSFER_ERROR_INTERNAL_ERROR_1);
+        return;
+    }
+
+    delete _voidStorageItems[slot];
+    _voidStorageItems[slot] = NULL;
```
Comment (Appled) [inline src/world/Units/Players/Player.cpp:4357]: Do not use NULL, always nullptr

File: src/world/Units/Players/Player.cpp [hunk 12]
```diff
@@ -4071,3 +4290,177 @@ bool Player::isOnGMTargetList(uint32_t guid) const
...
+    std::swap(_voidStorageItems[newSlot], _voidStorageItems[oldSlot]);
+    return true;
+}
+
+VoidStorageItem* Player::getVoidStorageItem(uint8_t slot) const
+{
+    if (slot >= VOID_STORAGE_MAX_SLOT)
+    {
+        GetSession()->sendVoidStorageTransferResult(VOID_TRANSFER_ERROR_INTERNAL_ERROR_1);
+        return NULL;
```
Comment (Appled) [inline src/world/Units/Players/Player.cpp:4374]: Do not use NULL, always nullptr

File: src/world/Units/Players/Player.cpp [hunk 13]
```diff
@@ -4071,3 +4290,177 @@ bool Player::isOnGMTargetList(uint32_t guid) const
...
+    for (uint8_t i = 0; i < VOID_STORAGE_MAX_SLOT; ++i)
+    {
+        if (_voidStorageItems[i] && _voidStorageItems[i]->itemId == id)
+        {
+            slot = i;
+            return _voidStorageItems[i];
+        }
+    }
+
+    return NULL;
```
Comment (Appled) [inline src/world/Units/Players/Player.cpp:4391]: Do not use NULL, always nullptr

**Review by Appled — CHANGES_REQUESTED** [review]

**Review by Appled — APPROVED** [review]

**Pull request merged**